vsphere/resource_vsphere_host.go

```diff
@@ -0,0 +1,711 @@
+	client := meta.(*VSphereClient).vimClient
+
+	// First let's establish where we are and where we want to go
+	var desiredConnectionState bool
+	if d.HasChange("connected") {
+		_, newVal := d.GetChange("connected")
+		desiredConnectionState = newVal.(bool)
+	} else {
+		desiredConnectionState = d.Get("connected").(bool)
+	}
+
+	hostID := d.Id()
+	hostObject, err := hostsystem.FromID(client, hostID)
+	if err != nil {
+		return err
+	}
+
+	actualConnectionState, err := hostsystem.GetConnectionState(hostObject)
+	if err != nil {
+		return err
+	}
+
+	// Have there been any changes that warrant a reconnect?
+	reconnect := false
+	connectionKeys := []string{"hostname", "username", "password", "thumbprint"}
+	for _, k := range connectionKeys {
+		if d.HasChange(k) {
+			reconnect = true
+			break
+		}
+	}
+
+	// Decide if we're going to reconnect or not
+	reconnectNeeded, err := shouldReconnect(d, meta, actualConnectionState, desiredConnectionState, reconnect)
+	if err != nil {
+		return err
+	}
+
+	switch reconnectNeeded {
+	case 1:
+		err := resourceVSphereHostReconnect(d, meta)
+		if err != nil {
+			return err
+		}
+	case -1:
+		err := resourceVSphereHostDisconnect(d, meta)
+		if err != nil {
+			return err
+		}
+	case 0:
+		break
+	}
+
+	mutableKeys := map[string]func(*schema.ResourceData, interface{}, interface{}, interface{}) error{
+		"license":     resourceVSphereHostUpdateLicense,
+		"cluster":     resourceVSphereHostUpdateCluster,
+		"maintenance": resourceVSphereHostUpdateMaintenanceMode,
+		"lockdown":    resourceVSphereHostUpdateLockdownMode,
+		"thumbprint":  resourceVSphereHostUpdateThumbprint,
+	}
+	for k, v := range mutableKeys {
+		log.Printf("[DEBUG] Checking if key %s changed", k)
+		if !d.HasChange(k) {
+			continue
+		}
+		log.Printf("[DEBUG] Key %s has change, processing", k)
+		old, newVal := d.GetChange(k)
+		err := v(d, meta, old, newVal)
+		if err != nil {
+			return fmt.Errorf("error while updating %s: %s", k, err)
+		}
+	}
+	return resourceVsphereHostRead(d, meta)
+}
+
+func resourceVsphereHostDelete(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*VSphereClient).vimClient
+	hostID := d.Id()
+
+	hs, err := hostsystem.FromID(client, hostID)
+	if err != nil {
+		return err
+	}
+
+	connectionState, err := hostsystem.GetConnectionState(hs)
+	if err != nil {
+		return err
+	}
+
+	if connectionState != types.HostSystemConnectionStateDisconnected {
+		// We cannot put a disconnected server in maintenance mode.
+		err = hostsystem.EnterMaintenanceMode(hs, int(defaultAPITimeout), true)
+		if err != nil {
+			return fmt.Errorf("error while putting host to maintenance mode: %s", err.Error())
+		}
+	}
+
+	hostProps, err := hostsystem.Properties(hs)
+	if err != nil {
+		log.Printf("Error while retrieving properties fort host %s. Error: %s", hostID, err)
+		return err
+	}
+
+	// If this is a standalone host we need to destroy the ComputeResource object
```

Comment:
What are the two options here? What does it mean to be a `ComputeResource` vs the alternative? This is a question for my understanding more than an question of the implementation or criticism.

Reply:
The parent for the hostsystem is either a ComputeResource (which means this is a standalone host we're deleting) or a ClusterComputeResouce (which means the host is a member of a cluster). 

Calling HostSystem.Destroy() on a "standalone" host (i.e a host that's not part of a vsphere cluster) then the task returns with a failure. In order to remove the HostSystem, you need to destroy the parent ComputeResource, and then vsphere will remove the HostSystem on its own.

Reply:
:+1: